framework/src/org/checkerframework/common/value/ValueTransfer.java

```diff
@@ -555,56 +628,131 @@ private AnnotationMirror createNumberAnnotationMirror(List<Number> values) {
                 }
             }
         }
+
+        // Refine the values of the operands in the store based on the result of the comparison.
+
+        List<Number> thenLeftVals = new ArrayList<>();
+        List<Number> elseLeftVals = new ArrayList<>();
+
+        List<Number> thenRightVals = new ArrayList<>();
+        List<Number> elseRightVals = new ArrayList<>();
+
+        // Because resultValues contains the (boolean) results of each comparison (i.e. the product of the number
+        // of elements in each list), we have to iterate through parts of the list at a time in order to find all
+        // the results of comparisons involving a particular value in the left list.
+        for (int i = 0; i < lefts.size(); i++) {
+            for (int j = 0; j < rights.size(); j++) {
+                if (resultValues.get(j + i * rights.size())) {
```

Comment:
It still isn't clear why this is the correct index.  Can you populate these lists in the for loop above?